syntax = "proto3";

service ObjectDetectionService {
    rpc Train (TrainRequest) returns (TrainResponse);
    rpc Inference (InferenceRequest) returns (InferenceResponse);
    rpc DeployBestModel (DeployBestModelRequest) returns (DeployBestModelResponse);
}

message TrainRequest {
    string run_name = 1;
    int32 epochs = 2;
    int32 batch_size = 3;
    bool plots = 4;
}

message TrainResponse {
    string message = 1;
}

message InferenceRequest {
    string image_path = 1;
}

message InferenceResponse {
    repeated ObjectDetectionResult results = 1;
}

message DeployBestModelRequest {}

message DeployBestModelResponse {
    string message = 1;
}

message ObjectDetectionResult {
    repeated BBox xyxy = 1;
    repeated float confidence = 2;
    bytes visualization_image = 3;
}

message BBox {
    int32 x1 = 1;
    int32 y1 = 2;
    int32 x2 = 3;
    int32 y2 = 4;
}
